### PR TITLE
[java] Fix #537: UnnecessaryParentheses fails to detect obvious scenario

### LINKED ISSUE #537: [java] UnnecessaryParentheses fails to detect obvious scenario
**Rule Set:**
java-controversial/UnnecessaryParentheses

**Description:**
The rule is currently too naive, only looking for parenthesis around a single term, but fails for several other trivial scenarios.

I'm not sure this will be achievable in XPath, we may have to convert the rule to a Java one....

**Code Sample demonstrating the issue:**

```java
public class Foo {
  private String getData(final String header) {
    String result = ("[" + header + ":" + Utils.decode(header) + "]");
    return result;
  }
}
```

**Running PMD through:** *Any*

### PR DESCRIPTION
**Pull request by oowekyala**
Fix #537. Not trying to be exhaustive, but I added a couple of tests for other scenarios too

### CODE REVIEW

**Comment by oowekyala** [comment]
Well since that rule is a duplicate maybe we should just deprecate it in favour of the unnecessary/UselessParentheses, which seems to consider more cases (arithmetic expressions, etc.). We could move the tests too if that makes sense

**Comment by jsotuyod** [comment]
@oowekyala absolutely! move tests, remove the rule, add a rename filter on compatibility ruleset factory.

File: pmd-core/src/main/java/net/sourceforge/pmd/RuleSetFactoryCompatibility.java [hunk 1]
```diff
@@ -67,6 +67,11 @@ public RuleSetFactoryCompatibility() {
         // unused/UnusedModifier moved AND renamed, order is important!
         addFilterRuleMoved("java", "unusedcode", "unnecessary", "UnusedModifier");
         addFilterRuleRenamed("java", "unnecessary", "UnusedModifier", "UnnecessaryModifier");
+
+        // PMD 6.0.0
+        addFilterRuleMoved("java", "controversial", "unnecessary", "UnnecessaryParentheses");
+        addFilterRuleMoved("java", "unnecessary", "UnnecessaryParentheses", "UselessParentheses");
```
Comment (jsotuyod) [inline pmd-core/src/main/java/net/sourceforge/pmd/RuleSetFactoryCompatibility.java]: I'm not sure we need this rename... I see no issue with the old name. It was actually consistent with the `java-unnecessary/UnnecessaryModifier` rule. Did you have any concerns with it?

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/unnecessary/xml/UselessParentheses.xml [hunk 2]
```diff
@@ -370,4 +370,81 @@ public class Useless {
 }
         ]]></code>
     </test-code>
+
+    <test-code>
```
Comment (jsotuyod) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/unnecessary/xml/UselessParentheses.xml]: this case is already in lines 99-110 of this same file

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/controversial/xml/UnnecessaryParentheses.xml [hunk 3]
```diff
@@ -1,88 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
-<test-data>
-    <test-code>
```
Comment (jsotuyod) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/controversial/xml/UnnecessaryParentheses.xml]: I don't see this case moved to the other test file

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/controversial/xml/UnnecessaryParentheses.xml [hunk 4]
```diff
@@ -1,88 +0,0 @@
...
-        <code><![CDATA[
-        public class Foo {
-          int bar() {
-            String a = ((String) "foo");
-          }
-        }
-        ]]></code>
-    </test-code>
-
-    <test-code>
```
Comment (jsotuyod) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/controversial/xml/UnnecessaryParentheses.xml]: I don't see this case moved to the other test file

File: pmd-core/src/main/java/net/sourceforge/pmd/RuleSetFactoryCompatibility.java (hunk 1 above)
Comment (jsotuyod) [inline pmd-core/src/main/java/net/sourceforge/pmd/RuleSetFactoryCompatibility.java]: this one should be `addFilterRuleRenamed`, isn't it?

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/controversial/xml/UnnecessaryParentheses.xml (hunk 3 above)
Comment (oowekyala) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/controversial/xml/UnnecessaryParentheses.xml]: There's already [this one](https://github.com/pmd/pmd/pull/554/files/69b32d4c0027885d1458f7a0bd9e55e37a2b3f52..2d5f1355c2fde0ed3eb12c99def8168ec4294c14#diff-35a3759c1754926dbc7d99dc1743033bR418) so I thought I could remove it

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/controversial/xml/UnnecessaryParentheses.xml (hunk 4 above)
Comment (oowekyala) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/controversial/xml/UnnecessaryParentheses.xml]: Right

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/unnecessary/xml/UselessParentheses.xml (hunk 2 above)
Comment (oowekyala) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/unnecessary/xml/UselessParentheses.xml]: Good catch!

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/controversial/xml/UnnecessaryParentheses.xml (hunk 3 above)
Comment (jsotuyod) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/controversial/xml/UnnecessaryParentheses.xml]: Ok! assignment and return statements are now treated equally, so I guess that's alright!

File: pmd-core/src/main/java/net/sourceforge/pmd/RuleSetFactoryCompatibility.java (hunk 1 above)
Comment (oowekyala) [inline pmd-core/src/main/java/net/sourceforge/pmd/RuleSetFactoryCompatibility.java]: Damn copy paste

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/unnecessary/xml/UselessParentheses.xml (hunk 2 above)
Comment (jsotuyod) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/unnecessary/xml/UselessParentheses.xml]: should still remove this case

Comment (oowekyala) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/unnecessary/xml/UselessParentheses.xml]: Hum no I already removed it, but for some reason your comment is on another test case now

Comment (oowekyala) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/unnecessary/xml/UselessParentheses.xml]: You can see it when viewing "changes from all commits"

Comment (jsotuyod) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/unnecessary/xml/UselessParentheses.xml]: No it's not... this is still the same test-case as https://github.com/oowekyala/pmd/blob/2d5f1355c2fde0ed3eb12c99def8168ec4294c14/pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/unnecessary/xml/UselessParentheses.xml#L100-L111

Comment (jsotuyod) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/unnecessary/xml/UselessParentheses.xml]: Oh, Github doing weird shit on me again. Nevermind

**Comment by jsotuyod** [comment]
Merged! Amazing!

**Pull request merged**